### PR TITLE
Publicly expose ability to register an error type on a mutation field

### PR DESCRIPTION
**Pull request by benmccallum**
Allows libraries to register their own error types in type interceptors (if they can tap in early enough)

### CODE REVIEW

**Comment by benmccallum** [comment]
This might've been what you were alluding to @michaelstaib when we last spoke. Exposing this to libraries.

Assuming it is, just need to address the todo in there and then this could be merged. I don't think I'll back port this to v12.

File: src/HotChocolate/Core/src/Types.Mutations/Extensions/ObjectFieldDefinitionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if (errorType is null)
+        {
+            throw new ArgumentNullException(nameof(errorType));
+        }
+
+        if (descriptorContext.ContextData[MutationContextDataKeys.Options] == null)
+        {
+            throw new Exception("Mutation convention isn't turned on");
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/Extensions/ObjectFieldDefinitionExtensions.cs:44]: This should be a SchemaException

**Comment by michaelstaib** [comment]
Can you add a test that shows this in action?

**Comment by benmccallum** [comment]
Thanks for the review. Will get back to this when I can

File: src/HotChocolate/Core/src/Types.Mutations/Extensions/ObjectFieldDefinitionExtensions.cs (hunk 1 above)
Comment (benmccallum) [reply src/HotChocolate/Core/src/Types.Mutations/Extensions/ObjectFieldDefinitionExtensions.cs]: Should be OK now

**Comment by benmccallum** [comment]
Hey @michaelstaib, test added and schema exception used (with a resource-file based msg).

File: src/HotChocolate/Core/test/Types.Mutations.Tests/__snapshots__/AnnotationBasedMutations.SimpleMutation_Add_Error_Via_Type_Interceptor.graphql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+type DoSomethingPayload {
+  string: String
+  errors: [DoSomethingError!]
+}
+
+type OutOfMemoryError implements Error {
+  message: String!
+}
```
Comment (benmccallum) [inline src/HotChocolate/Core/test/Types.Mutations.Tests/__snapshots__/AnnotationBasedMutations.SimpleMutation_Add_Error_Via_Type_Interceptor.graphql:16]: This one is added by a type inteceptor

File: src/HotChocolate/Core/src/Types/Properties/TypeResources.Designer.cs
Comment (benmccallum) [inline src/HotChocolate/Core/src/Types/Properties/TypeResources.Designer.cs:1]: I'm not sure why the resource file gen has so many differences... I've had this before on this repo and it also results in `TypeResources.___AppliedDirective_Description` becoming `TypeResources.__AppliedDirective_Description` (one less preceding `_`). You might need to pull and regen this file. Though it does look like my version correctly encodes `<` amd `>` ;)

**Review by michaelstaib — APPROVED** [review]

**Comment by michaelstaib** [comment]
Thanks for this one mate :)

**Comment by glen-84** [comment]
Was this merged into the correct branch?

**Pull request merged**